Logging without Limits guide (#6079)
DataDog/documentation: content/en/logs/guide/identify-and-exclude-noisy-logs.md

---
title: Identify and Exclude Noisy Logs
kind: guide
disable_toc: true
further_reading:
- link: "logs/explorer"
  tag: "Documentation"
  text: "Learn more about Log Explorer"
- link: "logs/explorer/patterns/"
  tag: "Documentation"
  text: "Learn about log patterns"
- link: "logs/live_tail"
  tag: "Documentation"
  text: "Explore Live Tail"
- link: "logs/logs_to_metrics"
  tag: "Documentation"
  text: "Learn how to generate metrics from ingested logs"
---

## Overview

Datadog’s [Logging without Limits][1]&trade; removes logging limitations by decoupling log ingestion and indexing to give flexibility around log management. High volume log indexing is helpful when monitoring every aspect of a service, but it can be a distraction when serious problems occur.

> For example, you may not find it beneficial to monitor every 200 response code log from a web server as you know this is not relevant for troubleshooting. It’s more beneficial to filter these logs so when an issue occurs you can quickly drill down to the source of the problem and resolve it.

Reviewer: ```suggestion
For example, you may not find it beneficial to monitor every 200 response code log from a web server, as you know this is not relevant for troubleshooting. It’s more beneficial to filter these logs so when an issue occurs, you can quickly drill down to the source of the problem and resolve it.
```

> For example, you may not find it beneficial to monitor every 200 response code log from a web server, as you know this is not relevant for troubleshooting. It’s more beneficial to filter these logs so when an issue occurs, you can quickly drill down to the source of the problem and resolve it.

Reviewer 2: ```suggestion
For example, you may not find it beneficial to monitor every 200 response code log from a web server, as it's not always relevant for troubleshooting issues. While 4xx and 5xx response code are critical to investigate problems. 
However, exhaustive KPI (i.e. Number of requests, ratio of errors, number of visitors...) should be computed upon all logs (including the 200 response code logs).
```

> This guide uses Logging without Limits&trade; to identify a noisy logging service and status, exclude irrelevant status logs, and set custom metrics from the excluded logs to continue to track data over time.

Reviewer 2: ```suggestion
This guide uses Logging without Limits&trade; to identify a noisy logging service and status, exclude irrelevant status logs, and set custom metrics from the excluded logs to continue to track KPI over time.
```

> ## 1. Identify a noisy logging service and status

Reviewer 2: We should add screenshots to this section to better demonstrate how to setup a log analytics + the final result of this process.

> Your noisiest logging service will contain several logs, some of which may be irrelevant for troubleshooting. By identifying your noisiest logging service, you can quickly find the information you need to filter irrelevant logs by status.

Reviewer: ```suggestion
Your noisiest logging service contains several logs, some of which may be irrelevant for troubleshooting. By identifying your noisiest logging service, you can quickly find the information you need to filter irrelevant logs by status.
```

**To identify your noisiest logging service and status**:

1. In Log Explorer, select graph view located next to the search bar. 
2. Below the search bar, set `count` * group by `service` and `limit` to top 10.
3. Select `Top List` from the dropdown menu next to `hide controls`.
4. Click on the first listed service and select `search for` in the populated menu. This generates a search, which is visible in the search bar above, based on your `service` facet.
5. Switch group by `service` to group by `status`. This generates a top statuses list for your service.

> 6. Click on the first listed status and select `search for` in the populated menu. This will add your `status` facet to the search.

Reviewer: ```suggestion
6. Click on the first listed status and select `search for` in the populated menu. This adds your `status` facet to the search.
```

> **Note**: These steps are applicable to any high volume logging query to generate a top list. You can group by any facet, such as `host` or `client name` vs. `service` or `status`.

Reviewer: ```suggestion
**Note**: These steps are applicable to any high volume logging query to generate a top list. You can group by any facet, such as `host` or `client name` versus `service` or `status`.
```

## 2. Identify noisy logs with log patterns
Switch to the [patterns][2] view, located next to the graph view, to view the patterns associated with your service and status.

> {{< img src="logs/guide/log-patterns.png" alt="Log Patterns View" responsive="true" style="width:75%;">}}

Reviewer 2: The context selected in the previous step is missing in this screenshot.

The patterns view is helpful when identifying and filtering noisy patterns. It shows the number of logs matching a pattern, split by service and status. Click on the first pattern to view a detailed log of events relating to your status. A contextual panel will populate with information about your noisiest status pattern.

> ## 3. Set an exclusion filter for noisy logs

Reviewer 2: Proposal:

```suggestion
## 3. Exclude noisy logs with an exclusion filter
```

the idea is to be consistent with previous titles.

> The selected pattern contextual panel lists every instance (event) of a log pattern and creates a custom parsing rule based on the selected pattern. This parsing rule is required when creating an [exclusion filter][3]. Exclusion filters remove noisy logs from your Log Explorer, but these logs still have the ability to be tracked with custom metrics.

Reviewer 2: ```suggestion
The selected pattern contextual panel lists every instance (event) of a log pattern and creates a custom search query based on the selected pattern. Use this query in an exclusion filter to remove those logs from your index. Note that excluded logs are still available in the [live tail view](/logs/live_tail/) and can be [sent to log archives](/logs/archives/), or used to [generate metrics](/logs/logs_to_metrics/).
```

**To create an exclusion filter for your top log pattern**:

1. Click on your top pattern from the pattern view list.

> 2. Click the parsing rule button in the top right corner to automatically generate a parsing rule associated with this pattern.

Reviewer 2: We should add a screenshot here to show where it happens in the UI

> 2. Click the parsing rule button in the top right corner to automatically generate a parsing rule associated with this pattern.

Reviewer 2: ```suggestion
2. Click the *View All* button in the top right corner to automatically generate the search query associated with this pattern.
```

> 3. Copy the parsing rule and navigate to the Configuration page under Logs in the main left hand navigation.

Reviewer 2: ```suggestion
3. Copy the search query and navigate to the [Index Configuration page](https://app.datadoghq.com/logs/pipelines/indexes) under *Logs* -> *Configuration* in the main left hand navigation.
```

> 3. Copy the parsing rule and navigate to the Configuration page under Logs in the main left hand navigation.

Reviewer 2: ```suggestion
3. Copy the search query and navigate to the Configuration page under Logs in the main left hand navigation.
```

4. Select indexes and click on your associated index. This will populate the option to add an exclusion filter.
5. Select “Add an Exclusion Filter”.

> 6. Input the filter name, define the exclusion query by pasting the parsing rule you copied, and set an exclusion percentage.

Reviewer 2: We should add a final screenshot of the exclusion filter configuration.

> 6. Input the filter name, define the exclusion query by pasting the parsing rule you copied, and set an exclusion percentage.

Reviewer 2: ```suggestion
6. Input the filter name, define the exclusion query by pasting the search query you copied, and set an exclusion percentage.
```

> You can also generate the parsing rule by clicking the **View All** button in the pattern contextual panel and selecting the **</>** option next to the search bar.

Reviewer 2: This is the preferred way to get the search query associated with a pattern working with an exclusion filter.

> You can also generate the parsing rule by selecting the `View All` button in the pattern contextual panel and selecting the **</>** option next to the search bar.

Reviewer: ```suggestion
You can also generate the parsing rule by clicking the **View All** button in the pattern contextual panel and selecting the **</>** option next to the search bar.
```

Exclusion filters can be disabled at any time by toggling the disable option to the right of the filter. They can also be modified and removed by hovering over the filter and selecting the edit or delete option.

**Note**: If a log matches several exclusion filters, only the first exclusion filter rule is applied. A log is not sampled or excluded multiple times by different exclusion filters.

## 4. Generate metrics to monitor excluded logs

> When a log pattern is excluded from Log Explorer, it still is available to view in [Live Tail][4]. However, you can continue to monitor excluded logs over time to generate KPIs from logs at ingest level by creating a new [log-based metric][5]. You can also create an [anomaly detection][6] monitor to notify you of any important changes to an excluded status.

Reviewer 2: ```suggestion
When a subset of logs is excluded from your index, those logs are still available to view in [Live Tail][4]. 
Moreover, you can generate KPIs from those excluded logs over time at ingest level by creating a new [log-based metric][5]. This metric can then be used for example to create an [anomaly detection][6] monitor to notify you of any unexpected changes.
```

### Add a new log-based metric

> To generate a new log-based metric based on your log pattern, go to the Logs Configuration page of your Datadog account and select the [Generate Metrics tab][7], then the New Metric+ button.

Reviewer: ```suggestion
To generate a new log-based metric based on your log pattern, go to the Logs Configuration page of your Datadog account and select the [Generate Metrics tab][7], then the **New Metric+** button.
```

> 1. Under Define Query, select the **</>** option and input your parsing rule.

Reviewer 2: ```suggestion
1. Under Define Query, select the **</>** option and input your search query.
```

> 2. Select the field you would like to track: Select * to generate a count of all logs matching your query or enter a log attribute (e.g., @network.bytes_written) to aggregate a numeric value and create its corresponding count, min, max, sum, and avg aggregated metrics.

Reviewer: ```suggestion
2. Select the field you would like to track: Select `*` to generate a count of all logs matching your query or enter a log attribute (e.g., `@network.bytes_written`) to aggregate a numeric value and create its corresponding count, min, max, sum, and avg aggregated metrics.
```

> 3. Add dimensions to group by: Select log attributes or tag keys to apply to the generated log-based metric to transform them into [tags][8] following the <KEY>:<VALUE> format. Log-based metrics are considered [custom metrics][9]. Avoid grouping by unbounded or extremely high cardinality attributes like timestamps, user IDs, request IDs, or session IDs to avert impacting your billing.

Reviewer 2: ```suggestion
3. Add dimensions to group by: Select log attributes or tag keys to apply to the generated log-based metric to transform them into [tags][8] following the `<KEY>:<VALUE>` format. Log-based metrics are considered [custom metrics][9]. Avoid grouping by unbounded or extremely high cardinality attributes like timestamps, user IDs, request IDs, or session IDs to avert impacting your billing.
```

> 4. Name your metric: Log-based metric names must follow the [naming metric convention][10].

Reviewer 2: We should add a screenshot of a full configuration.

### Create an anomaly detection monitor
{{< img src="logs/guide/anomaly-detection-monitor.gif" alt="Anomaly Detection Monitor" responsive="true" style="width:75%;">}}

Anomaly detection is an algorithmic feature that identifies when a metric is behaving differently than it has in the past. Creating an anomaly detection monitor for your excluded logs will alert you of any changes based on your set alert conditions.

1. To set an anomaly detection monitor, go to [Monitors -> New Monitor -> Anomaly][11].

> 2. Set the detection method as anomaly detection.

Reviewer 2: This should be the default, no need to explicitly set it.

> 3. Define your metric by selecting the **</>** option and paste your parsing rule into the search bar.

Reviewer 2: This is for a log monitor not an anomaly monitor, the step should be "enter the log-based metric you defined in the previous section"

Reviewer 2: We should maybe describe here what is happening in the gif (split down by service, focus on a given status..)

4. Set the alert conditions and add any additional information needed to alert yourself and/or your team of what’s happening.
5. Save the monitor.

> When an anomaly is detected, an alert will appear in [Monitors -> Trigger Monitors][12].

Reviewer: ```suggestion
When an anomaly is detected, an alert appears in [Monitors -> Trigger Monitors][12].
```

---
title: Identify and Exclude Noisy Logs
kind: guide
disable_toc: true
further_reading:
- link: "logs/explorer"
  tag: "Documentation"
  text: "Learn more about Log Explorer"
- link: "logs/explorer/patterns/"
  tag: "Documentation"
  text: "Learn about log patterns"
- link: "logs/live_tail"
  tag: "Documentation"
  text: "Explore Live Tail"
- link: "logs/logs_to_metrics"
  tag: "Documentation"
  text: "Learn how to generate metrics from ingested logs"
---

## Overview

Datadog’s [Logging without Limits][1]&trade; removes logging limitations by decoupling log ingestion and indexing to give flexibility around log management. High volume log indexing is helpful when monitoring every aspect of a service, but it can be a distraction when serious problems occur.

For example, you may not find it beneficial to monitor every 200 response code log from a web server, as it's not always relevant for troubleshooting issues. While 4xx and 5xx response code are critical to investigate problems. 

> However, exhaustive KPI (i.e. Number of requests, ratio of errors, number of visitors...) should be computed upon all logs (including the 200 response code logs).

Reviewer 2: Not really sure if we can do that ? 

```suggestion
However, exhaustive KPIs (i.e. Number of requests, ratio of errors, number of visitors...) should be computed upon all logs (including the 200 response code logs).
```

> This guide uses Logging without Limits&trade; to identify a noisy logging service and status, exclude irrelevant status logs, and set custom metrics from the excluded logs to continue to track KPI over time.

Reviewer 2: ```suggestion
This guide uses Logging without Limits&trade; to identify a noisy logging service status, exclude irrelevant status logs, and set custom metrics from the excluded logs to continue to track KPI over time.
```

> ## 1. Identify a noisy logging service and status

Reviewer 2: ```suggestion
## 1. Identify a noisy logging service status
```

Your noisiest logging service contains several logs, some of which may be irrelevant for troubleshooting. By identifying your noisiest logging service, you can quickly find the information you need to filter irrelevant logs by status.

{{< img src="logs/guide/logging-without-limits.gif" alt="Log Patterns View" responsive="true" style="width:100%;">}}

> **To identify your noisiest logging service and status**:

Reviewer 2: ```suggestion
**To identify your noisiest logging service status**:
```

> 1. In Log Explorer, select graph view located next to the search bar. 

Reviewer 2: We should maybe add a link to the in-app log explorer page here.

## 2. Identify noisy logs with log patterns
Switch to the [patterns][2] view, located next to the graph view, to view the patterns associated with your service and status.

{{< img src="logs/guide/patterns.png" alt="Log Patterns View" responsive="true" style="width:100%;">}}

The patterns view is helpful when identifying and filtering noisy patterns. It shows the number of logs matching a pattern, split by service and status. Click on the first pattern to view a detailed log of events relating to your status. A contextual panel will populate with information about your noisiest status pattern.

## 3. Exclude noisy logs with an exclusion filter

The selected pattern contextual panel lists every instance (event) of a log pattern and creates a custom search query based on the selected pattern. Use this query in an exclusion filter to remove those logs from your index. Note that excluded logs are still available in the [live tail view][3] and can be [sent to log archives][4], or used to [generate metrics][5].

**To create an exclusion filter for your top log pattern**:

> {{< img src="logs/guide/exclusion-filter.png" alt="Log Patterns View" responsive="true" style="width:100%;">}}

Reviewer 2: The image should go after the list of instructions i think, since it represents the final result of those instructions.

1. Click on your top pattern from the pattern view list.
2. Click the *View All* button in the top right corner to automatically generate the search query associated with this pattern.
3. Select the `</>` option to the right of the search query and copy the search query.
4. Navigate to the Configuration page under Logs in the main left hand navigation. Select indexes and click on your associated index. This will populate the option to add an exclusion filter.
5. Select “Add an Exclusion Filter”.
6. Input the filter name, define the exclusion query by pasting the search query you copied, and set an exclusion percentage.

Exclusion filters can be disabled at any time by toggling the disable option to the right of the filter. They can also be modified and removed by hovering over the filter and selecting the edit or delete option.

**Note**: If a log matches several exclusion filters, only the first exclusion filter rule is applied. A log is not sampled or excluded multiple times by different exclusion filters.

## 4. Generate metrics to monitor excluded logs

> {{< img src="logs/guide/generate-metric.png" alt="Log Patterns View" responsive="true" style="width:75%;">}}

Reviewer 2: This image should go after the instructions to create the log-based metric as the final result of the instructions

When a subset of logs is excluded from your index, those logs are still available to view in [Live Tail][6]. Moreover, you can generate KPIs from those excluded logs over time at ingest level by creating a new [log-based metric][7]. This metric can then be used for example to create an [anomaly detection][8] monitor to notify you of any unexpected changes.

### Add a new log-based metric
To generate a new log-based metric based on your log pattern, go to the Logs Configuration page of your Datadog account and select the [Generate Metrics tab][9], then the **New Metric+** button.

1. Under Define Query, input the search query you got from the pattern in step 2.

> 2. Select the field you would like to track: Select `*` to generate a count of all logs matching your query or enter a log attribute (e.g., `@network.bytes_written`) to aggregate a numeric value and create its corresponding count, min, max, sum, and avg aggregated metrics.

Reviewer 2: ```suggestion
2. Select the field you would like to track: Select `*` to generate a count of all logs matching your query or enter a measure (e.g., `@network.bytes_written`) to aggregate a numeric value and create its corresponding count, min, max, sum, and avg aggregated metrics.
```

we might also link to the measure doc here

3. Add dimensions to group by: Select log attributes or tag keys to apply to the generated log-based metric to transform them into [tags][10] following the `<KEY>:<VALUE>` format. Log-based metrics are considered [custom metrics][11]. Avoid grouping by unbounded or extremely high cardinality attributes like timestamps, user IDs, request IDs, or session IDs to avert impacting your billing.
4. Name your metric: Log-based metric names must follow the [naming metric convention][12].

### Create an anomaly detection monitor
Anomaly detection is an algorithmic feature that identifies when a metric is behaving differently than it has in the past. Creating an anomaly detection monitor for your excluded logs will alert you of any changes based on your set alert conditions.

> {{< img src="logs/guide/anomaly-alert.png" alt="Log Patterns View" responsive="true" style="width:75%;">}}

Reviewer 2: The notification image should be put after the last sentence of this section.

In order to be more relevant? Could we create a "fake metric" `webstore.ok` with an anomaly monitor that drops studently to better emphase the anomaly detection?

Author: Updated with all suggestions. Waiting on the fake metric, will add once finalized.
